python/private/py_console_script_gen.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2023 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""entry_point generator."""
+
+from __future__ import annotations
+
+import argparse
+import configparser
+import pathlib
+import re
+import sys
+import textwrap
+
+_ENTRY_POINTS_TXT = "entry_points.txt"
+
+# If we create a `py_console_script_binary` in the main workspace, then the first entry in the `sys.path` will allow
+# the Python scripts to load things from outside the `.runfiles` directory. This means that programs like `pylint`
+# and `flake8` get confused trying to load from `pylint` when the `py_console_script_binary` is named `pylint`, hence
+# we remove the first entry point for this reason.
+#
+# This would not happen if we created an console_script binary in the root of an external repository, e.g.
+# `@pypi_pylint//` because that path is treated differently.
+
+_TEMPLATE = """\
+import sys
+
+# See @rules_python//python/private:py_console_script_gen.py for explanation
+if ".runfiles" not in sys.path[0]:
+    sys.path = sys.path[1:]
+
+try:
+    from {module} import {attr}
+except ImportError:
+    entries = "\\n".join(sys.path)
+    print("Printing sys.path entries for easier debugging:", file=sys.stderr)
+    print(f"sys.path is:\\n{{entries}}", file=sys.stderr)
+    raise
+
+if __name__ == "__main__":
+    sys.exit({entry_point}())
+"""
+
+
+class EntryPointsParser(configparser.ConfigParser):
+    """A class handling entry_points.txt
+
+    See https://packaging.python.org/en/latest/specifications/entry-points/
+    """
+
+    optionxform = staticmethod(str)
+
+
+def _normalize(name):
+    """
+    Taken from https://packaging.python.org/en/latest/specifications/name-normalization/
+    and tweaked a bit
+    """
+    return re.sub(r"[-_.]+", "_", name).lower()
+
+
+def _guess_entry_point(guess: str, console_scripts: dict[string, string]) -> str | None:
```

Comment:
I'm -1 on this normalized guessing; leaving a comment to flag this; let's resolve on the issue.

Reply:
Removed normalizing.